### ISSUE TITLE
Auto infer output type coming from `np.vectorize`

### ISSUE THREAD

**Issue by gab23r** [pola-rs/polars#12607]
### Description

When I use `np.vectorize`, I can give the type of the output using `otypes`.
It would be nice if polars could use this inforamtion (if given) to cast to the right type.

# here I give  otypes=[int] but I get a Utf8 dtype
```
import polars as pl
import numpy as np


levenshtein = pl.reduce(np.vectorize(Levenshtein.distance, otypes=[int]), [pl.col('A'), pl.col('B')]).explode()
pl.LazyFrame({'A': list('abc'), 'B': list('edc')}).with_columns(levenshtein = levenshtein).schema
# OrderedDict([('A', Utf8), ('B', Utf8), ('levenshtein', Utf8)])

```

**Comment by alexander-beedie** [comment]
I can see how to improve this... we should add a new (optional) `dtype` argument for `reduce`, `fold` & co so the caller can specify the output dtype more precisely (and if we can infer it on the Python side - like we could here- and the caller _doesn't_ set it, we can do a better job) 🤔

**Comment by deanm0000** [comment]
On a semi related note @ion-elgreco made [this extension](https://github.com/ion-elgreco/polars-distance) which has levenshtein distance.

**Comment by deanm0000** [comment]
@alexander-beedie as long as you're contemplating changes to reduce, is there anything to be done for [this](https://github.com/pola-rs/polars/issues/10512) issue?

**Comment by gab23r** [comment]
Yes, I was using @ion-elgreco 's plugin (and had a X2 speedup). But this extension is not yet available with the new polars version

**Comment by ion-elgreco** [comment]
@gab23r I'll push a new release tomorrow!

**Comment by gab23r** [comment]
I think that this issue clutter the issue board more than anything, I am not even sure than polars should do more here.
